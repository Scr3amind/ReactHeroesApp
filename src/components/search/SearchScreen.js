import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useLocation } from 'react-router';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../heroes/HeroCard';

export const SearchScreen = ({history}) => {

    const location = useLocation();

    const {q = ''} = queryString.parse(location.search);

    const [formValues, handleInputChange] = useForm({searchText: q})
    const {searchText} = formValues;
    const heroesFiltered = useMemo(() => getHeroesByName(q), [q])

    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${searchText}`);
    }

    return (
        <div className="flex flex-col items-center  ">
            <h1 className="font-semibold text-2xl">Search</h1>
            
            <form className="text-white" onSubmit={handleSearch}>
                <input
                    className="h-14 w-80 outline-none text-xl bg-gray-700 opacity-80 shadow-lg rounded-xl px-2" 
                    type="text"   
                    placeholder="Find your hero"
                    name="searchText" 
                    autoComplete="off"
                    value={searchText}
                    onChange={handleInputChange}
                />                    
                <button 
                    className="ml-5 bg-gray-800 text-white px-5 py-2 rounded-xl shadow-lg"
                    type="submit"
                >
                    Search
                </button>
            </form>
            <div className="flex gap-4 justify-center my-10 flex-wrap ">
                {
                    (q !== '' && heroesFiltered.length === 0) && 
                    <div className="bg-red-100 border-l-4 border-red-500 text-red-700 p-4 shadow-lg rounded-md">
                        <p className="font-bold">No Result</p>
                        <p>There is no hero {q}.</p>
                    </div>

                }
                {heroesFiltered.map(hero => <HeroCard key={hero.id} {...hero}/>)}
            </div>
            
        </div>
    )
}
