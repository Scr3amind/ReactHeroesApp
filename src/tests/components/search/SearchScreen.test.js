import { mount } from 'enzyme';
import { MemoryRouter, Route, Router } from 'react-router';
import { SearchScreen } from '../../../components/search/SearchScreen';
import '../../../setupTest';

describe('Pruebas en <SearchScreen />', () => {
    
    test('debe de mostrarse correctamente con valores por defecto', () => {
        
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect(wrapper).toMatchSnapshot();

    })

    test('debe de mostrar a batman y el input con el valor del queryString', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman');
        expect(wrapper).toMatchSnapshot();
    })

    test('debe de mostrar un error si no se encuentra el hero', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman145']}>
                <Route path="/search" component={SearchScreen}/>
            </MemoryRouter>
        );

        expect(wrapper.html().includes('No Result')).toBe(true);
    })

    test('debe llamar el push del history ', () => {
        const history = {
            push: jest.fn(),
            listen: jest.fn()
        };

        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>

                <Route path="/search" component={() => <SearchScreen history={history} />}/>
        
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change',{
            target:{
                name: 'searchText',
                value: 'batman'
            }
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });

        expect(history.push).toHaveBeenCalledWith('?q=batman');


    })
    
    
    
    

})
