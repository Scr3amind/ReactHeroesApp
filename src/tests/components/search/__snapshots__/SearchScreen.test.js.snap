// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchScreen /> debe de mostrar a batman y el input con el valor del queryString 1`] = `
<div
  className="flex flex-col items-center  "
>
  <h1
    className="font-semibold text-2xl"
  >
    Search
  </h1>
  <form
    className="text-white"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="h-14 w-80 outline-none text-xl bg-gray-700 opacity-80 shadow-lg rounded-xl px-2"
      name="searchText"
      onChange={[Function]}
      placeholder="Find your hero"
      type="text"
      value="batman"
    />
    <button
      className="ml-5 bg-gray-800 text-white px-5 py-2 rounded-xl shadow-lg"
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="flex gap-4 justify-center my-10 flex-wrap "
  >
    <div
      className="flex items-center justify-around h-52 w-72 shadow-lg rounded-xl bg-gradient-to-r from-gray-100 to-gray-200 animate__animated animate__fadeIn"
    >
      <div
        className="flex justify-center w-1/2"
      >
        <img
          alt="Batman"
          className="rounded-md shadow-lg h-48"
          src="./assets/heroes/dc-batman.jpg"
        />
      </div>
      <div
        className="flex flex-col justify-around h-full w-1/2 text-center"
      >
        <p
          className="font-bold"
        >
          Batman
        </p>
        <p>
          Bruce Wayne
        </p>
        <p
          className="text-xs"
        >
          Detective Comics #27
        </p>
        <a
          className="text-blue-800"
          href="/hero/dc-batman"
          onClick={[Function]}
        >
          Ver mas...
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Pruebas en <SearchScreen /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div
  className="flex flex-col items-center  "
>
  <h1
    className="font-semibold text-2xl"
  >
    Search
  </h1>
  <form
    className="text-white"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="h-14 w-80 outline-none text-xl bg-gray-700 opacity-80 shadow-lg rounded-xl px-2"
      name="searchText"
      onChange={[Function]}
      placeholder="Find your hero"
      type="text"
      value=""
    />
    <button
      className="ml-5 bg-gray-800 text-white px-5 py-2 rounded-xl shadow-lg"
      type="submit"
    >
      Search
    </button>
  </form>
  <div
    className="flex gap-4 justify-center my-10 flex-wrap "
  />
</div>
`;
